Fix tensor registration to work with coalescing collectives.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98643
* #98642
* #99765
* #99764
* __->__ #99763

We do it by making it possible to register multiple tensors for the same
worker and coordinate waiting/cleanup among them.

This ensures waiting on any number the output tensors will result in a
single stream sync. This simplifies codegen by inductor.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

kumpera committed:
Fix tensor registration to work with coalescing collectives.

We do it by making it possible to register multiple tensors for the same
worker and coordinate waiting/cleanup among them.

This ensures waiting on any number the output tensors will result in a
single stream sync. This simplifies codegen by inductor.

[ghstack-poisoned]

wanchaol commented on torch/distributed/_functional_collectives.py:
```diff
@@ -79,56 +78,72 @@ def _is_running_under_torch_deploy():
+    def cleanup(self):
+        for ptr in self.ptrs:
+            if data_ptr_to_work.get(ptr, None) == self:
+                del data_ptr_to_work[ptr]
```
when cleaning up, do we need to clean up the `self.ptrs` list too?

kumpera replied:
`self.ptrs` is a list of ints, so it's fine to leave it as is and we don't have to worry about cleanup running multiple times as we guard against that by checking the key in the dict.

wanchaol reviewed (approved):
lgtm, thanks for cleaning this up! Have one comment

kumpera commented:
@pytorchmergebot merge

kumpera commented:
@pytorchbot rebase

kumpera committed:
Update on "Fix tensor registration to work with coalescing collectives."


We do it by making it possible to register multiple tensors for the same
worker and coordinate waiting/cleanup among them.

This ensures waiting on any number the output tensors will result in a
single stream sync. This simplifies codegen by inductor.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

kumpera commented:
@pytorchmergebot merge -f "inductor crash unrelated"
